src/raster/buffer.rs

```diff
@@ -0,0 +1,305 @@
+    }
+
+    /// Destructures `self` into constituent parts.
+    pub fn into_shape_and_vec(self) -> ((usize, usize), Vec<T>) {
+        (self.shape, self.data)
+    }
+
+    /// Gets the 2-d shape of the buffer.
+    //
+    /// Returns `(cols, rows)`
+    ///
+    /// # Notes
+    /// The elements of `shape` are in reverse order from what is used in `ndarray`.
+    pub fn shape(&self) -> (usize, usize) {
+        self.shape
+    }
+
+    /// Get a slice over the buffer contents.
+    pub fn data(&self) -> &[T] {
+        self.data.as_slice()
+    }
+
+    /// Get a mutable slice over the buffer contents.
+    pub fn data_mut(&mut self) -> &mut [T] {
+        self.data.as_mut_slice()
+    }
+
+    /// Get the number of elements in the buffer
+    pub fn len(&self) -> usize {
+        self.data.len()
+    }
+
+    /// Determine if the buffer has no elements.
+    pub fn is_empty(&self) -> bool {
+        self.data.is_empty()
+    }
+
+    #[cfg(feature = "ndarray")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "ndarray")))]
+    /// Convert `self` into an [`ndarray::Array2<T>`].
+    pub fn to_array(self) -> crate::errors::Result<Array2<T>> {
+        // Array2 shape is (rows, cols) and Buffer shape is (cols in x-axis, rows in y-axis)
+        Ok(Array2::from_shape_vec(
+            (self.shape.1, self.shape.0),
+            self.data,
+        )?)
+    }
+
+    #[inline]
+    fn vec_index_for(&self, coord: (usize, usize)) -> usize {
+        if coord.0 >= self.shape.0 {
+            panic!(
+                "index out of bounds: buffer has {} columns but row {} was requested",
+                self.shape.0, coord.0
+            );
+        }
+        if coord.1 >= self.shape.1 {
+            panic!(
+                "index out of bounds: buffer has {} rows but row {} was requested",
+                self.shape.1, coord.1
+            );
+        }
+        coord.0 * self.shape.0 + coord.1
+    }
+}
+
+impl<T: GdalType> Index<(usize, usize)> for Buffer<T> {
+    type Output = T;
+    fn index(&self, index: (usize, usize)) -> &Self::Output {
+        &self.data[self.vec_index_for(index)]
+    }
+}
+
+impl<T: GdalType> IndexMut<(usize, usize)> for Buffer<T> {
+    fn index_mut(&mut self, index: (usize, usize)) -> &mut Self::Output {
+        let idx = self.vec_index_for(index);
+        &mut self.data[idx]
+    }
+}
+
+impl<T: GdalType> IntoIterator for Buffer<T> {
+    type Item = T;
+    type IntoIter = IntoIter<T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.data.into_iter()
+    }
+}
+
+impl<'a, T: GdalType> IntoIterator for &'a Buffer<T> {
+    type Item = &'a T;
+    type IntoIter = Iter<'a, T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.data.iter()
+    }
+}
+
+impl<'a, T: GdalType> IntoIterator for &'a mut Buffer<T> {
+    type Item = &'a mut T;
+    type IntoIter = IterMut<'a, T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.data.iter_mut()
+    }
+}
+
+pub type ByteBuffer = Buffer<u8>;
```

Comment:
Why do we even have this? :confused:

Reply:
no clue. Chesterton's Fence.

Reply:
I almost removed it (only in one test) until I saw it was `pub`.